engines/asylum/puzzles/pipes.cpp

```diff
@@ -0,0 +1,652 @@
+		connect(_nextConnector);
+}
+
+void Connector::turn(bool updpos) {
+	if (updpos)
+		*_position = (*_position == 8) ? 1 : *_position << 1;
+	BinNum newState = BinNum(_state >> 1 | (_state & 1) << 3);
+
+	uint32 delta = _state ^ newState;
+	uint32 newIndex[2], oldIndex[2];
+
+	if (delta == kBinNum1111) {
+		if (newState == kBinNum0101) {
+			newIndex[0] = 0;
+			newIndex[1] = 2;
+			oldIndex[0] = 1;
+			oldIndex[1] = 3;
+		} else {
+			newIndex[0] = 1;
+			newIndex[1] = 3;
+			oldIndex[0] = 0;
+			oldIndex[1] = 2;
+		}
+	} else {
+		newIndex[0] = (uint32)Common::intLog2(newState & delta);
+		oldIndex[0] = (uint32)Common::intLog2(_state & delta);
+	}
+
+	for (uint32 i = 0; i < (uint32)(delta == kBinNum1111 ? 2 : 1); ++i) {
+		if (_nodes[oldIndex[i]]) {
+			_nodes[oldIndex[i]]->disconnect(this);
+			_connectedNodes.remove(_nodes[oldIndex[i]]);
+		}
+
+		if (_nodes[newIndex[i]]) {
+			_nodes[newIndex[i]]->connect(this);
+			_connectedNodes.push_back(_nodes[newIndex[i]]);
+		}
+	}
+
+	_state = newState;
+
+	if (_nextConnector) {
+		if (_isConnected) {
+			if (!(_nextConnectorPosition & _state))
+				disconnect(_nextConnector);
+		} else if (_nextConnectorPosition & _state && _nextConnector->isReadyForConnection()) {
+			connect(_nextConnector);
+		}
+	}
+}
+
+void Connector::connect(Connector *connector) {
+	for (Common::List<Peephole *>::iterator iter = _connectedNodes.begin(); iter != _connectedNodes.end(); ++iter) {
+		(*iter)->connect(connector);
+		connector->_connectedNodes.push_back(*iter);
+	}
+
+	for (Common::List<Peephole *>::iterator iter = connector->_connectedNodes.begin(); iter != connector->_connectedNodes.end(); ++iter) {
+		(*iter)->connect(this);
+		_connectedNodes.push_back(*iter);
+	}
+
+	_isConnected = connector->_isConnected = true;
+
+}
+
+void Connector::disconnect(Connector *connector) {
+	uint32 i;
+	Common::List<Common::List<Peephole *>::iterator> markedForDeletion;
+	bool flag;
+
+	for (i = 0; i < 4; ++i)
+		if (_nodes[i]) {
+			_nodes[i]->disconnect(connector);
+			connector->_connectedNodes.remove(_nodes[i]);
+		}
+
+	for (Common::List<Peephole *>::iterator iter = _connectedNodes.begin(); iter != _connectedNodes.end(); ++iter) {
+		flag = true;
+
+		for (i = 0; i < 4; ++i) {
+			if (*iter == _nodes[i]) {
+				flag = false;
+				break;
+			}
+		}
+
+		if (flag)
+			markedForDeletion.push_back(iter);
+	}
+
+	for (Common::List<Common::List<Peephole *>::iterator>::iterator iter1 = markedForDeletion.begin(); iter1 != markedForDeletion.end(); ++iter1) {
+		(*(*iter1))->disconnect(this);
+		_connectedNodes.remove(*(*iter1));
+	}
+
+	_isConnected = connector->_isConnected = false;
+}
+
+//////////////////////////////////////////////////////////////////////////
+// Spider
+//////////////////////////////////////////////////////////////////////////
+Spider::Spider(const Common::Rect &rect, Common::String id) {
+	_boundingBox = rect;
+	_rnd = new Common::RandomSource(Common::String("pipes_spider") + id);
```

Comment:
What is the point of creating yet another random source instead of using the global one from the engine level?